samples/js/photos/src/photo-grid.js

```diff
@@ -0,0 +1,355 @@
+// @flow
+
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+import React from 'react';
+import ReactDOM from 'react-dom';
+import {
+  AsyncIterator,
+  notNull,
+} from '@attic/noms';
+import cache from './cache.js';
+import {searchToParams, paramsToSearch} from './dom.js';
+import Nav from './nav.js';
+import {default as Photo, createPhoto} from './photo.js';
+import type {PhotoSize, NomsPhoto} from './types.js';
+import {EmptyIterator} from './photo-set.js';
+import Viewport from './viewport.js';
+
+const maxPhotoHeight = 300;
+const timeGroupThresholdMs = 5000;
+const photosPerPage = 10;
+const photoSpacing = 5;
+const transitionDelay = '200ms';
+
+type PhotoCompProps = {
+  fullscreen: boolean,
+  gridHeight: number,
+  gridLeft: number,
+  gridSize: PhotoSize,
+  gridTop: number,
+  gridWidth: number,
+  nav: Nav,
+  photo: Photo,
+  url: string,
+  viewport: Viewport,
+}
+
+type PhotoCompState = {
+  size: PhotoSize,
+  sizeIsBest: boolean,
+  url: string,
+}
+
+class PhotoComponent extends React.Component<void, PhotoCompProps, PhotoCompState> {
+  state: PhotoCompState;
+  _parentTop: number;
+  _parentLeft: number;
+  _shouldTransition: boolean;
+
+  constructor(props: PhotoCompProps) {
+    super(props);
+    this.state = {size: props.gridSize, sizeIsBest: false, url: props.url};
+    this._parentTop = 0;
+    this._parentLeft = 0;
+    this._shouldTransition = true;
+  }
+
+  componentWillUpdate() {
+    const rect = ReactDOM.findDOMNode(this).parentElement.getBoundingClientRect();
+    this._parentTop = rect.top;
+    this._parentLeft = rect.left;
+  }
+
+  render(): React.Element<*> {
+    const {fullscreen, photo, viewport} = this.props;
+    const {sizeIsBest, url} = this.state;
+
+    let clipStyle, imgStyle, overlay;
+    if (fullscreen) {
+      clipStyle = this._getClipFullscreenStyle();
+      imgStyle = this._getImgFullscreenStyle();
+      overlay = <div style={{
+        animation: `fadeIn ${transitionDelay}`,
+        backgroundColor: 'black',
+        position: 'fixed',
+        top: 0, right: 0, bottom: 0, left: 0,
+        zIndex: 1, // above photo grid, below fullscreen img because it's before in the DOM
+      }}/>;
+      if (!sizeIsBest) {
+        // The fullscreen image is low-res, fetch the hi-res version.
+        // disabled to avoid glaring errors, though it's still not perfect - if an image loads in
+        // less than transitionDelay then it will look a bit wonky.
+        // TODO: Also fade in the hi-res image.
+        const [bestSize, bestUrl] = photo.getBestSize(viewport.clientWidth, viewport.clientHeight);
+        cache(bestUrl).then(() => {
```

Comment:
Yes, load into the cache. You could probably structure it to immediately zoom the low-res version out, but that would be more transform wrangling, and this looks reasonable.